external-crates/move/move-abstract-stack/src/unit_tests/mod.rs

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::*;
+
+#[test]
+fn test_empty_stack() {
+    let mut empty = AbsStack::<usize>::new();
+    assert!(empty.is_empty());
+
+    // pop on empty stack
+    assert_eq!(empty.pop(), Err(AbsStackError::EmptyStack));
+    assert_eq!(empty.pop_any_n(nonzero(1)), Err(AbsStackError::EmptyStack));
+    assert_eq!(
+        empty.pop_any_n(nonzero(100)),
+        Err(AbsStackError::EmptyStack)
+    );
+    assert_eq!(empty.pop_eq_n(nonzero(12)), Err(AbsStackError::EmptyStack));
+    assert_eq!(empty.pop_eq_n(nonzero(112)), Err(AbsStackError::EmptyStack));
+
+    assert!(empty.is_empty());
+}
+
+#[test]
+fn test_simple_push_pop() {
+    let mut s = AbsStack::new();
+    s.push(1).unwrap();
+    assert!(!s.is_empty());
+    assert_eq!(s.len(), 1);
+    assert_eq!(s.pop(), Ok(1));
+    assert!(s.is_empty());
+    assert_eq!(s.len(), 0);
+
+    s.push(1).unwrap();
+    s.push(2).unwrap();
+    s.push(3).unwrap();
+    assert!(!s.is_empty());
+    assert_eq!(s.len(), 3);
+    assert_eq!(s.pop(), Ok(3));
+    assert_eq!(s.pop(), Ok(2));
+    assert_eq!(s.pop(), Ok(1));
+    assert!(s.is_empty());
+    assert_eq!(s.len(), 0);
+
+    s.push_n(1, 1).unwrap();
+    s.push_n(2, 2).unwrap();
+    s.push_n(3, 3).unwrap();
+    assert!(!s.is_empty());
+    assert_eq!(s.len(), 6);
+    assert_eq!(s.pop(), Ok(3));
+    assert_eq!(s.pop(), Ok(3));
+    assert_eq!(s.pop(), Ok(3));
+    assert_eq!(s.pop(), Ok(2));
+    assert_eq!(s.pop(), Ok(2));
+    assert_eq!(s.pop(), Ok(1));
+    assert!(s.is_empty());
+    assert_eq!(s.len(), 0);
+
+    s.push_n(1, 1).unwrap();
+    s.push_n(2, 2).unwrap();
+    s.push_n(3, 3).unwrap();
+    assert!(!s.is_empty());
+    assert_eq!(s.len(), 6);
+    assert_eq!(s.pop_eq_n(nonzero(3)), Ok(3));
+    assert_eq!(s.pop_eq_n(nonzero(2)), Ok(2));
+    assert_eq!(s.pop_eq_n(nonzero(1)), Ok(1));
+    assert!(s.is_empty());
+    assert_eq!(s.len(), 0);
+
+    s.push(1).unwrap();
+    s.push(2).unwrap();
+    s.push(2).unwrap();
+    s.push(3).unwrap();
+    s.push(3).unwrap();
+    s.push(3).unwrap();
+    assert!(!s.is_empty());
+    assert_eq!(s.len(), 6);
```

Comment:
wonder if there was an easy way in this test to get into the internal of the stack and check that "compaction" had happened....

Reply:
It's a real good question. I'll poke at it